internal/hncconfig/reconciler.go

```diff
@@ -145,38 +137,32 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 // make sure there's no dup and the types exist. Update the type set with GR to
 // GVK mappings.
 func (r *Reconciler) reconcileTypes(inst *api.HNCConfiguration) error {
-	// Get all resources for all groups.
-	allRes, err := getAPIGroupResources(r.Manager.GetConfig())
-	if err != nil {
-		r.Log.Error(err, "while trying to get all resources")
-		return err
-	}
-
 	// Overwrite the type set each time. Initialize them with the enforced types.
 	r.activeGVKMode = gr2gvkMode{}
 	r.activeGR = gvk2gr{}
-	if err := r.ensureEnforcedTypes(inst, allRes); err != nil {
+	if err := r.ensureEnforcedTypes(inst); err != nil {
 		// Early exit if any enforced types are not found for some reason to retry.
 		return err
 	}
 
 	// Add all valid configurations from user-configured types.
-	r.reconcileConfigTypes(inst, allRes)
+	r.reconcileConfigTypes(inst)
 	return nil
 }
 
 // ensureEnforcedTypes ensures HNC enforced types 'roles' and 'rolebindings' are
 // in the type set. Return error to retry (if any) since they are enforced types.
-func (r *Reconciler) ensureEnforcedTypes(inst *api.HNCConfiguration, allRes []*restmapper.APIGroupResources) error {
+func (r *Reconciler) ensureEnforcedTypes(inst *api.HNCConfiguration) error {
 	for _, t := range api.EnforcedTypes {
 		gr := schema.GroupResource{Group: t.Group, Resource: t.Resource}
 
 		// Look if the resource exists in the API server.
-		gvk, err := gvkForGR(gr, allRes)
+		gvk, err := r.ResourceMapper.NamespacedKindFor(gr)
 		if err != nil {
-			// If the type is not found, log error and write conditions and return the
+			// Error looking up enforced types. Log error, write conditions and return the
 			// error for a retry.
 			r.Log.Error(err, "while trying to reconcile the enforced resource", "resource", gr)
+			// TODO(adrianludwin): Should we rather use api.ReasonUnknown here?
 			r.writeCondition(inst, api.ConditionBadTypeConfiguration, api.ReasonResourceNotFound, err.Error())
```

Comment:
Yeah probably a good idea, you can go ahead and change that. Or maybe factor out the code at line 203 and use it here as well. Thanks!

Reply:
Good suggesting to align this with the other use of the function. I'll have to think a bit how to do it...

Reply:
I have added a suggestion. Please let me know what you think!